### PR TITLE
[serving] Check available memory before loading model

### PR DESCRIPTION
**Pull request by frankfliu**
## Description ##

Brief description of what this PR is about

- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

File: serving/docs/modes.md [hunk 1]
```diff
@@ -65,6 +65,10 @@ and python handler functions.
 - `gpu.maxWorkers`: Maximum number of workers for GPU.
 - `cpu.minWorkers`: Minimum number of workers for CPU.
 - `cpu.maxWorkers`: Maximum number of workers for CPU.
+- `required_memory_mb`: Specify the required memory (CPU and GPU) in MB to load the model.
```
Comment (siddvenk) [inline serving/docs/modes.md:68]: For Python based engines, are we assuming that this accounts for the number of mpi workers? Or should we account for mpi workers and parallel_loading in the estimate calculation?

File: wlm/src/main/java/ai/djl/serving/wlm/ModelInfo.java [hunk 2]
```diff
@@ -621,6 +632,80 @@ private void configPerModelSettings(Path modelDir) throws IOException {
...
+        long requiredMemory = intValue(prop, "required_memory_mb", -1) * 1024L * 1024;
+        WlmConfigManager wlmc = WlmConfigManager.getInstance();
+        int defMemory = wlmc.getReservedMemoryMb();
+        long reservedMemory = intValue(prop, "reserved_memory_mb", defMemory) * 1024L * 1024;
+        if (requiredMemory <= 0) {
+            try (Stream<Path> walk = Files.walk(modelDir)) {
+                // estimate the memory to be 1.2x of file size
+                requiredMemory =
+                        walk.filter(Files::isRegularFile).mapToLong(ModelInfo::getFileSize).sum();
+                requiredMemory = requiredMemory * 12 / 10;
```
Comment (siddvenk) [inline wlm/src/main/java/ai/djl/serving/wlm/ModelInfo.java:646]: Is 1.2x a valid multiplier for all models? I'm ok with this for now, but I think we determined this from our experiments with DS based models, so not sure how accurate it is for other models (as long as it doesn't severely underestimate the required memory it should be ok).

File: wlm/src/main/java/ai/djl/serving/wlm/ModelInfo.java [hunk 3]
```diff
@@ -206,7 +217,7 @@ public void load(Device device) throws ModelException, IOException {
 
             ZooModel<I, O> m = builder.build().loadModel();
             if (criteria != null) {
-                // TODO: use has to manually configure batchifer if using dynamic batch
+                // TODO: user has to manually configure birdwatcher if using dynamic batch
```
Comment (siddvenk) [inline wlm/src/main/java/ai/djl/serving/wlm/ModelInfo.java:220]: what is birdwatcher?

Comment (frankfliu) [reply wlm/src/main/java/ai/djl/serving/wlm/ModelInfo.java]: fixed

**Review by siddvenk — APPROVED** [review]

**Pull request merged**